### PR TITLE
new method to fetch ghcup yaml

### PR DESCRIPTION
**Pull request by PhotonQuantum**
- query yaml file from latest tag to avoid fetching unreleased files
- mirror all past yaml files, removing additional_yamls arg
- downgrade config version mismatch to warning

### CODE REVIEW

File: src/ghcup/utils.rs [hunk 1]
```diff
@@ -1,34 +1,177 @@
...
+            .get(format!(
+                "https://{}/api/v4/projects/{}/repository/tree",
+                config.host,
+                urlencoding::encode(&*config.repo)
             ))
+            .query(&[("per_page", config.pagination), ("page", page)])
+            .query(&[("ref", commit.clone())])
+            .send()
+            .await
+            .map_err(|_| Error::ProcessError(String::from("unable to list files")))?
```
Comment (skyzh) [inline src/ghcup/utils.rs:150]: `ProcessError` is for spawning process errors. Use another error type for this.

File: src/ghcup/utils.rs [hunk 2]
```diff
@@ -1,34 +1,177 @@
...
+    pub fn version(&self) -> Version {
+        self.version
+    }
+}
+
+pub async fn fetch_last_tag(client: &Client, config: &GhcupRepoConfig) -> Result<String> {
+    let req = client.get(format!(
+        "https://{}/api/v4/projects/{}/repository/tags",
+        config.host,
+        urlencoding::encode(&*config.repo)
```
Comment (skyzh) [inline src/ghcup/utils.rs:117]: Why do we need to encode? Isn't the `{}` be replaced with something like `user/repo`?

**Review by skyzh — COMMENTED** [review]
Rest LGTM

File: src/ghcup/utils.rs (hunk 2 above)
Comment (PhotonQuantum) [reply src/ghcup/utils.rs:117]: https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding

File: src/ghcup/yaml.rs [hunk 3]
```diff
@@ -1,27 +1,21 @@
+use std::collections::HashMap;
+
 use async_trait::async_trait;
+use itertools::Itertools;
 use slog::info;
-use structopt::StructOpt;
-use url::Url;
 
 use crate::common::{Mission, SnapshotConfig, TransferURL};
 use crate::error::{Error, Result};
```
Comment (skyzh) [inline src/ghcup/yaml.rs:8]: unused import `Error`

**Review by skyzh — COMMENTED** [review]
I'll test the new version and LGTM this weekend :)

**Review by skyzh — APPROVED** [review]
I've tested and updated the files on our mirror site. If ghcup could work correctly, you may merge this PR and release a new version yourself.

**Pull request merged**